cmake/FlatccGenerateSources.cmake

```diff
@@ -0,0 +1,86 @@
+ cmake_minimum_required(VERSION 3.5)
+
+# Use the following function to generate C source files from flatbuffer definition files:
+#
+# flatcc_generate_sources(GENERATED_SOURCE_DIRECTORY <directory where to write source files>
+#                         GENERATE_BUILDER
+#                         GENERATE_VERIFIER
+#                         EXPECTED_OUTPUT_FILES <list of files that flatcc is supposed to generate>
+#                         DEFINITION_FILES <list of flatbuffer definition files (.fbs)>
+# )
+#
+# GENERATE_BUILDER and GENERATE_VERIFIER are boolean options. When specified they will instruct
+# flatcc to generate builder / verifier source code.
+#
+# With cross-compiling you should provide the directory where the flatcc compiler executable is located
+# in environment variable FLATCC_BUILD_BIN_PATH. If you use Conan and add flatcc as a build requirement
+# this will be done automatically.
+
+
+function(flatcc_generate_sources)
+
+    # parse function arguments
+    set(OUTPREFIX "FLATCC") #variables created by 'cmake_parse_arguments' will be prefixed with this
+    set(NO_VAL_ARGS GENERATE_BUILDER GENERATE_VERIFIER)
+    set(SINGLE_VAL_ARGS GENERATED_SOURCE_DIRECTORY)
+    set(MULTI_VAL_ARGS DEFINITION_FILES EXPECTED_OUTPUT_FILES CC_OPTIONS)
+
+    cmake_parse_arguments(${OUTPREFIX}
+                          "${NO_VAL_ARGS}"
+                          "${SINGLE_VAL_ARGS}"
+                          "${MULTI_VAL_ARGS}"
+                          ${ARGN}
+    )
+    if (GENERATED_SOURCE_DIRECTORY IN_LIST FLATCC_KEYWORDS_MISSING_VALUES)
+        message(FATAL_ERROR "No directory provided after GENERATED_SOURCE_DIRECTORY keyword")
+    endif()
+    if (NOT FLATCC_GENERATED_SOURCE_DIRECTORY)
+        set(FLATCC_GENERATED_SOURCE_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR})
+    endif()
+    message(STATUS "Flatcc sources will be generated in directory ${FLATCC_GENERATED_SOURCE_DIRECTORY}")
+
+    if (FLATCC_GENERATE_BUILDER)
+        list(APPEND FLATCC_CC_OPTIONS --builder)
+    endif()
+    if (FLATCC_GENERATE_VERIFIER)
+        list(APPEND FLATCC_CC_OPTIONS --verifier)
+    endif()
+
+    if (FLATCC_DEFINITION_FILES)
+        if (NOT EXISTS ${FLATCC_GENERATED_SOURCE_DIRECTORY})
+            file(MAKE_DIRECTORY ${FLATCC_GENERATED_SOURCE_DIRECTORY})
+        endif()
+
+        message(VERBOSE "Executing command ${FLATCC_COMPILER} ${FLATCC_CC_OPTIONS} -o ${FLATCC_GENERATED_SOURCE_DIRECTORY} ${FLATCC_DEFINITION_FILES}")
+        add_custom_command(OUTPUT ${FLATCC_EXPECTED_OUTPUT_FILES}
```

Comment:
This needs `DEPENDS` for dependency tracking of the input flatbuffer files.

Reply:
Without reading carefully and from memory, just adding DEPENDS will break Ninja build because CMake doesn't handle dependecies properly, even if it works with Make. If you look at some of the examples that build with CMAKE, these are rather clumsy because of this.

Reply:
But without DEPENDS, the target sources don't get regenerated when the input sources get modified?

Reply:
[The `add_custom_command` documentation](https://cmake.org/cmake/help/latest/command/add_custom_command.html) does mention `ninja` with multiple arguments, but not with `DEPENDS` btw.

Reply:
I don't know because I am not a CMake expert. But I do know that I have spent several days trying to make it work (or something similar). Another who believed he could fix it had to admit he couldn't.

Reply:
I think DEPENDS is ok, I'm really don't recall the problem exactly. Here is the the monster test build:

https://github.com/dvidelabs/flatcc/blob/master/test/monster_test/CMakeLists.txt

Reply:
The problem I was having was that test cases cannot run properly with dependencies. I think I just disabled it with CMake as driver and ran tests with shell scripts.

Reply:
I'm looking into it now, I'll open a pr for installing `flatcc-config.cmake`.

This allows your users to use flatcc by doing `find_package(flatcc REQUIRED)` and call flatcc using `flatcc::flatcc`.

Reply:
It looks like on every `make`, all tests are re-built.
This is because the dependency tree is not correct.
I'll look into that too.

Reply:
@wdobbe
You can then re-use/rebase your `FlatccGenerateSources.cmake` script on my pr.

Reply:
OK, here is an old discussion about broken dependencies - I haven't reread it, but linked here for reference:

https://github.com/dvidelabs/flatcc/pull/24#issuecomment-266718814

Reply:
> I'm looking into it now, I'll open a pr for installing `flatcc-config.cmake`.
> 
> This allows your users to use flatcc by doing `find_package(flatcc REQUIRED)` and call flatcc using `flatcc::flatcc`.

Yes, that is a good idea. I have plenty examples of how to generate the config files (correctly) so if you want me to do it, let me know.